File: dotnet-impl/main/net/terria/registry/terraform/ModuleController.cs
using Microsoft.AspNetCore.Mvc;

namespace net.terria.registry.terraform
{
    [ApiController]
    [Route("v1/modules")]
    public class ModuleController : ControllerBase
    {
        private readonly RegistryService _registryService;
        private const string DefaultRegistry = "registry.terraform.io";

        public ModuleController(RegistryService registryService)
        {
            _registryService = registryService;
        }

        /// <summary>
        /// Terraform Registry Protocol: GET /v1/modules/{namespace}/{name}/versions
        /// Returns module versions from filesystem-backed index.
        /// </summary>
        [HttpGet("{namespace}/{name}/versions")]
        public IActionResult GetModuleVersions(string @namespace, string name)
        {
            var module = _registryService.GetModule(DefaultRegistry, @namespace, name);
            if (module == null) return NotFound();

            var versions = module.versions
                .OrderByDescending(v => v.Key)
                .Select(v => new
                {
                    version = v.Key,
                    systems = v.Value.systems
                });

            return Ok(new { versions });
        }

        /// <summary>
        /// Terraform Registry Protocol: GET /v1/modules/{namespace}/{name}/{system}/{version}/download
        /// Returns a JSON payload with the source archive download URL.
        /// </summary>
        [HttpGet("{namespace}/{name}/{system}/{version}/download")]
        public IActionResult DownloadModule(string @namespace, string name, string system, string version)
        {
            var module = _registryService.GetModule(DefaultRegistry, @namespace, name);
            if (module == null) return NotFound();

            if (!module.versions.ContainsKey(version)) return NotFound();

            var url = $"/storage/modules/{@namespace}/{name}/{version}/{system}/terraform-module-{name}_{system}_{version}.zip";
            return Ok(new { download_url = url });
        }
    }
}

File: java-impl/src/main/java/net/terria/registry/terraform/ModuleController.java
package net.terria.registry.terraform;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/v1/modules")
public class ModuleController {

    @Autowired
    private RegistryService registryService;
    
    private static final String DEFAULT_REGISTRY = "registry.terraform.io";

    /**
     * Terraform Registry Protocol: GET /v1/modules/{namespace}/{name}/versions
     * Returns module versions from filesystem-backed index.
     */
    @GetMapping("/{namespace}/{name}/versions")
    public ResponseEntity<?> getModuleVersions(@PathVariable String namespace,
                                               @PathVariable String name) {
        return registryService.getModule(DEFAULT_REGISTRY, namespace, name)
            .map(index -> {
                List<Map<String, Object>> versions = index.versions.entrySet().stream()
                    .map(entry -> Map.of(
                        "version", entry.getKey(),
                        "systems", entry.getValue().systems
                    ))
                    .collect(Collectors.toList());

                return ResponseEntity.ok(Map.of("versions", versions));
            })
            .orElse(ResponseEntity.notFound().build());
    }

    /**
     * Terraform Registry Protocol: GET /v1/modules/{namespace}/{name}/{system}/{version}/download
     * Returns a JSON payload with the source archive download URL.
     */
    @GetMapping("/{namespace}/{name}/{system}/{version}/download")
    public ResponseEntity<?> downloadModule(@PathVariable String namespace,
                                            @PathVariable String name,
                                            @PathVariable String system,
                                            @PathVariable String version) {
        return registryService.getModule(DEFAULT_REGISTRY, namespace, name)
            .map(index -> {
                if (!index.versions.containsKey(version)) {
                    return ResponseEntity.notFound().build();
                }

                String url = String.format("/storage/modules/%s/%s/%s/%s/terraform-module-%s_%s_%s.zip",
                    namespace, name, version, system, name, system, version);
                return ResponseEntity.ok(Map.of("download_url", url));
            })
            .orElse(ResponseEntity.notFound().build());
    }
}
